### ISSUE TITLE
[strict-boolean-expressions] allowNullableObject has no effect

### ISSUE THREAD

**Issue by floAtEbcont** [typescript-eslint/typescript-eslint#2155]
**Repro**

```JSON
{
  "rules": {
        "@typescript-eslint/strict-boolean-expressions": ["warn", {
          "allowNullableObject": true
        }]
  }
}
```

```TS
let myObject: {test: number} | null = null;
if (Math.random() > 0.5) {
  myObject = { test: 12 };
}
if (myObject) {
  console.log('null');
}
```

In my `tsconfig.json`:
```JSON
{
  "compilerOptions": {
    "strictNullChecks": false,
  }
}
```

**Expected Result**
No warning.

**Actual Result**
Warning thrown for `if (myObject) { // ...` , even though `myObject` is an object:

```5:5  warning  Unexpected object value in conditional. The condition is always true  @typescript-eslint/strict-boolean-expressions```

**Additional Info**
If I use `"strictNullChecks": true` the rule works as expected and no warning is thrown. I don't see any reason why this should depend on `strictNullChecks` though...?

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.1.0` (Also tried it with `3.1.1-alpha.0`)|
| `@typescript-eslint/parser`        | `3.1.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `7.1.0` |
| `node`                             | `13.8.0` |
| `npm`                              | `6.13.6` |

**Comment by bradzacher** [comment]
`"strictNullChecks": false,`

With this turned off, typescript essentially erases the `null` and `undefined` type from every union.
You cannot use this rule with that compiler option turned off.

**Comment by floAtEbcont** [comment]
I see - thank you very much for your detailed answer!

Maybe add a hint about `strictNullChecks` in [the docs](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/strict-boolean-expressions.md)? (Intuitively I would have guessed that `"allowNullableObject": true` just "turns off" that part of the rule completely, no matter what's set in tsconfig.)
